src/finn/custom_op/fpgadataflow/rtl/thresholding_rtl.py

```diff
@@ -0,0 +1,776 @@
+            # and IPI generation
+            "gen_top_module": ("s", False, ""),
+            # bias to be applied to outputs of the node
+            "activation_bias": ("i", False, 0),
+            # whether weights (thresholds) will be
+            # writable through an AXI-lite interface during runtime
+            # 1 for enabled, 0 for disabled.
+            "runtime_writeable_weights": ("i", False, 0, {0, 1}),
+            # memory depth triggers for threshold storage
+            "depth_trigger_uram": ("i", False, 0),
+            "depth_trigger_bram": ("i", False, 0),
+            # enable uniform thres optimization
+            # doesn't actually do anything yet, only
+            # for resource estimations
+            "uniform_thres": ("i", False, 0, {0, 1}),
+            # enable deep pipelining for easier timing closure
+            # setting to 0 may save some FFs but otherwise leave on
+            "deep_pipeline": ("i", False, 1, {0, 1}),
+        }
+        my_attrs.update(super().get_nodeattr_types())
+        return my_attrs
+
+    def get_pe_mem_geometries(self):
+        pe = self.get_nodeattr("PE")
+        wdt = self.get_weight_datatype()
+        wdt_bits = wdt.bitwidth()
+        odt = self.get_output_datatype()
+        odt_bits = odt.bitwidth()
+        t_channels = self.get_nodeattr("NumChannels")
+        cf = t_channels / pe
+        is_uniform = self.get_nodeattr("uniform_thres")
+        if is_uniform:
+            ret = [(odt_bits - x, cf * (2**x)) for x in range(1, odt_bits)]
+        else:
+            ret = [(wdt_bits, (cf) * 2**x) for x in range(odt_bits)]
+        return ret
+
+    def get_memory_estimate(self):
+        res_dict = {}
+        depth_trigger_bram = self.get_nodeattr("depth_trigger_bram")
+        depth_trigger_uram = self.get_nodeattr("depth_trigger_uram")
+        pe = self.get_nodeattr("PE")
+        ret = self.get_pe_mem_geometries()
+        for mem_cfg in ret:
+            (width, depth) = mem_cfg
+            primitives = mem_primitives_versal
+            if depth_trigger_bram != 0 or depth_trigger_uram != 0:
+                if depth >= depth_trigger_bram and depth < depth_trigger_uram:
+                    primitives = {k: v for (k, v) in mem_primitives_versal.items() if "BRAM" in k}
+                elif depth >= depth_trigger_uram:
+                    primitives = {k: v for (k, v) in mem_primitives_versal.items() if "URAM" in k}
+            alts = get_memutil_alternatives(mem_cfg, primitives)
+            primary_alt = alts[0]
+            res_type = primary_alt[0].split("_")[0]
+            res_count, eff, waste = primary_alt[1]
+            res_dict[res_type] = res_dict.get(res_type, 0) + pe * res_count
+        return res_dict
+
+    def calc_tmem(self):
+        """Calculates and returns TMEM."""
+        num_channels = self.get_nodeattr("NumChannels")
+        pe = self.get_nodeattr("PE")
+        return num_channels // pe
+
+    def make_shape_compatible_op(self, model):
```

Comment:
This functions can be removed and the functionality from the hw abstraction layer can be utilized. E.g. when the user calls `Thresholding_rtl.make_shape_compatible(model)` and the function is not implemented on this level, because of class inheritance `Thresholding.make_shape_compatible(model)` will be executed automatically. This allows us to implement functions that are backend independent in the HW abstraction layer.

Reply:
You can check which functions are implemented here: https://github.com/Xilinx/finn/blob/refactor/rtl_integration/src/finn/custom_op/fpgadataflow/thresholding.py#L66
And if the functions in this files are just copies, you can safely remove them.